utils/ipc.py

```diff
@@ -1,102 +1,157 @@
+from __future__ import annotations
+
 import asyncio
-import aiohttp
-import json
 import os
+
+from typing import Any, AsyncIterator, Callable, Coroutine, Dict, Optional, Set, TypedDict
+
+import aiohttp
+
 from discord.ext import ipc
-from typing import Any, AsyncGenerator, Callable, Optional, Union, Dict
+
+from utils.useful import print_exception
+
+
+class IPCData(TypedDict, total=False):
+    # requires python 3.11: https://www.python.org/dev/peps/pep-0655
+    # error: Dict[str, Any]
+    ...
+
+
+class _RecvMessagePayload(TypedDict):
+    endpoint: str
+    # requires python 3.11: https://www.python.org/dev/peps/pep-0655
+    # request_id: NotRequired[str]
+    response: IPCData
+
+
+class _SendMessagePayload(TypedDict):
+    endpoint: str
+    request_id: str
+    headers: Dict[str, Any]
+    data: Dict[str, Any]
+
+
+_HandlerType = Callable[[IPCData], Coroutine[Any, Any, Any]]
 
 
 class StellaClient(ipc.Client):
     def __init__(self, *args: Any, **kwargs: Any):
         super().__init__(*args, **kwargs)
         self.bot_id = kwargs.pop("bot_id", None)
-        self._listeners = {}
-        self.events = {}
-        self.connect = None
+        # callbacks are subscribed to event name and message id
+        self._callbacks: Dict[str, Dict[str, asyncio.Future[IPCData]]] = {}
+        # event handlers are subscribed to event name
+        self._event_handlers: Dict[str, Set[_HandlerType]] = {}
```

Comment:
I dont like _HandlerType is in a set, make it a list, I want order and I don't want much memory from storing callbacks